### ISSUE TITLE
Provide officials Lua API headers for LuaLS

### ISSUE THREAD

**Issue by AFCMS** [minetest/minetest#13926]
### Problem

Currently there are no advertised modern IDE experience for Minetest modding, which is quite frustrating for newcomers, especially when most developers for exemple in web development are used to TypeScript auto completion and linting in their code editors.

There is a Lua language server, [LuaLS](https://luals.github.io), which uses the open [Microsoft's Language Server Protocol](https://microsoft.github.io/language-server-protocol/) and have an good popularity, especially it's VSCode extension which have almost 1M downloads.

LSP is supported by a wide range of editors (some of them require plugins):
- VSCode and derivated
- Visual Studio
- Jetbrains Products (short term)
   - IntellIJ based already have support in early access program
   - Fleet have LSP support in it's roadmap
- Neovim
- Vim (not all versions)
- Emacs
- Sublime Text
- [And many more](https://microsoft.github.io/language-server-protocol/implementors/tools/)

Just using the language server can already provide a premium grade Lua programming experience, with syntax checking, type checking and hover documentation for builtin functions, but Minetest functions and objects do not get type checking and documentation. 

### Solutions

To get autocompletion and linting for Minetest API, users need to use definition files, which are Lua files with just global API functions and variables declared with documentation and type annotations in comments.

There were efforts made by many people including me and @lua-rocks:

- https://github.com/minetest-toolkit/minetest-lsp-api
- https://github.com/AFCMS/minetest_vscode_lua

These are mostly complete, but they have huge problems since they are not official.

They use different conventions especially for naming of classes. It doesn't matter if you just use it for autocompletion, but if you want to add typing annotations to your own code, it will lead to inconsistency if two contributors of your project uses different headers.

Having official API headers with documentation would serve as the reference for this language server / typing system.

What I suggest is:

- Make official API headers (two ways)
   - Generate them automatically from the builtin lua code + headers made by hand for the C++ defined API
   - Manually create them in a separate repository, can use some already done work as a starting point. Would require to manually keep sync of Minetest API changes (at least roughly), but do no loose the benefit of being the reference of API headers. Could also be tagged at each Minetest release to make managing multiple versions of installed headers easier.
- Provide official guide to using the language server on popular editors
   - Provide configuration file specially crafted for Minetest
   - Provide guide to use with VSCode, Jetbrains (in the future) and other popular editors among Minetest community
- IN THE FUTURE: I would like to go further and have official IDE integration (editors, linting of game files, translation editor, etc) as plugins for VSCode, Jetbrains, etc. Would really make Minetest development much more user friendly for beginners considering Godot have an official VSCode extension, Roblox have a complete IDE, Minecraft Bedrock have a VSCode / VS completion and linting support, etc.

### Alternatives

None

**Comment by AFCMS** [comment]
**Hey core developers :wave:**

I am _really_ interested into working on this. (and later on the IDE stuff)
But it require several things first:
- defining style of the headers
   - files organisation?
   - rules about the way of writing markdown help texts?
   - "class" names (most important)
- creating the GitHub repository
- writing already the release of headers together with the Minetest release process?

These things must be discussed by you before I can start working.

Feel free to reach me on GitHub or Discord if you want to discuss with me.

I can't wait to work on this! :grin:

**Comment by rubenwardy** [comment]
I support this because having a computer readable version of the Lua API has huge benefits. It can be used for Lua LS, but you could also use it to generate TypeScript definition files.

I don't have any preference on the format or organisation, as long as it's a standard way of doing it and not something custom

**Comment by sfan5** [comment]
how do the "lua headers" actually look? is there type info?

ideally we'd have some sort of machine-readable definition (not json) from which we can generate 1) lua_api.md 2) some sort of real HTML documentation 3) these Lua headers

**Comment by AFCMS** [comment]
I am unsure if we can really "generate" stuff like headers, since typing mechanism may be complex and largely differ between different systems. `lua_api.md` would be great but I don't really know since it uses a section design with various data, not just the API documentation.

I don't think we can avoid some duplication anyways.

My idea was to handle these headers somehow like the translations, with the feature freeze being the time to make sure every API feature is properly added.

**Comment by sfan5** [comment]
> I am unsure if we can really "generate" stuff like headers, since typing mechanism may be complex and largely differ between different systems.

True, if you wanted to try and map every last detail the complexity would explode.
But even if you still have some parts of the lua headers or lua_api.md that is hand-written because it can't be generated we should aim to have as much as possible in machine-readable form.

> My idea was to handle these headers somehow like the translations, with the feature freeze being the time to make sure every API feature is properly added.

problem situation: it is now the feature freeze and nobody wants to bother with the Lua header updates. What happens to them?

<hr />

Anyway I wrote up this Proof of Concept in an hour to demonstrate what I had on my mind: https://gist.github.com/sfan5/71715b190d06aaf81a49a3332d6d01d5

You put in a YAML specification of functions and it generates both the markdown and the LuaLS-compatible output you see there.

**Comment by AFCMS** [comment]
> True, if you wanted to try and map every last detail the complexity would explode. But even if you still have some parts of the lua headers or lua_api.md that is hand-written because it can't be generated we should aim to have as much as possible in machine-readable form.

The maintenance of headers like these is really not that hard, so IDK if its worth the additional complexity of an added generation system.
 
> problem situation: it is now the feature freeze and nobody wants to bother with the Lua header updates. What happens to them?

If its official headers and we have a guide to use them (and maybe IDE plugins that use them in the future) I think there will always be people to keep them updated with Minetest versions. At least 4 peoples have been working on non-official headers already. Using only Lua files means the maintenance cost is pretty minimal (almost copy paste of lua_api most of the time, with just a little bit more work in case of added classes or using generics).

> Anyway I wrote up this Proof of Concept in an hour to demonstrate what I had on my mind: https://gist.github.com/sfan5/71715b190d06aaf81a49a3332d6d01d5
> 
> You put in a YAML specification of functions and it generates both the markdown and the LuaLS-compatible output you see there.

The biggest problem with this approach IMO is the fact that declaring typing do not work the same depending of the typing system, so this format won't really help much more by itself than the Lua version if someone wants to build headers/types for another typing/language system. And then we lose some of the easiness of writing directly Lua code (no real autocompletion even if providing a JSON schema for the YAML files, not all features covered by the generation, etc).

**Comment by BuckarooBanzay** [comment]
tangentially related: https://github.com/lunarmodules/luacheck/pull/108 (luacheck builtin for minetest)

**Comment by Desour** [comment]
:+1: For machine-readable lua API documentation, and for generating `lua_api.md` from it.

This would also allow making things such as proper function search, clickable links, or tables of contents.
[LDoc](https://lunarmodules.github.io/ldoc/), for example, makes rather nice docs.

**Comment by AFCMS** [comment]
LuaLS can export documentation in Markdown (with a bit less features than LDoc), but it can also export as JSON so it can be easily read by another program without using a Lua parser. 

https://luals.github.io/wiki/export-docs

**Comment by appgurueu** [comment]
> LuaLS can export documentation in Markdown (with a bit less features than LDoc), but it can also export as JSON so it can be easily read by another program without using a Lua parser.

I would still prefer LDoc, though, especially as it has proper features for parsing C++ files. We want machine-readable, in-source Lua documentation, and LDoc is simply *the* tool for that.

It should be possible to generate whatever language server or linter files you want from there, though maybe it requires a bit of work.

**Comment by Saltane** [comment]
Currently learning Luanti modding, almost none of previous programming experience. Current documentation did not make it easy to know what value types can be put in tables and functions, and returned from functions. Had to do figure things out myself by logging, table dumping, and just trying until success.

Have been lately playing with Teal language, rewriting one of my mods to it. To make builder not scream at me, had to start writing Teal type definitions for Luanti myself. Current documentation, unfortunately, is not sufficient as a reference to create a proper typedef file.

Human readable .md file in @sfan5 's proof of concept looks very good to me. It would be sufficient to (mostly) eliminate problem described in first paragraph, and make problem in second paragraph much more easier and straightforward to solve.

I also have a suggestion: differentiate between numbers and integers.
Teal defines two types of numbers: `number` for any number, and `number`'s subtype `integer` for numbers without points. It is sometimes important to separate the two, as some functions and table fields don't operate with fractions. For example, item count in item stacks, node position, and table indices in lists.

**Comment by nauta-turbidus** [comment]
Separating `number` and `integer` in documentation is of course easy. In code (`type()` etc.), not necessarily.

Anyway, language server can go a long way in ensuring code quality etc., and certainly official headers like that would be helpful.

**Comment by appgurueu** [comment]
I think there is consensus that we should try to move to a machine-readable format for documentation, from which we can then generate typedefs as well as human-readable documentation. Additional supplementary documentation can go to https://docs.luanti.org/ or Markdown files in the repo.

We should try to come to an agreement on this format. Some options are LDoc, LuaLS, or a custom format. See e.g. LÖVR for an example of a custom Lua-based DSL format that works fine: https://github.com/bjornbytes/lovr-docs/blob/v0.18.0/api/lovr/math/Mat4/rotate.lua (rendered output: https://lovr.org/docs/Mat4:rotate).

**Comment by nauta-turbidus** [comment]
> I think there is consensus that we should try to move to a machine-readable format for documentation, from which we can then generate typedefs as well as human-readable documentation. Additional supplementary documentation can go to https://docs.luanti.org/ or Markdown files in the repo.
> 
> We should try to come to an agreement on this format. Some options are LDoc, LuaLS, or a custom format. See e.g. LÖVR for an example of a custom Lua-based DSL format that works fine: https://github.com/bjornbytes/lovr-docs/blob/v0.18.0/api/lovr/math/Mat4/rotate.lua (rendered output: https://lovr.org/docs/Mat4:rotate).

There is an effort going on at https://git.minetest.land/archie/luanti-api and in parallel elsewhere.

**Comment by appgurueu** [comment]
I'm aware of such efforts, they're a good thing for sure, but that doesn't mean we shouldn't pursue having a more useful source of documentation truth in the engine?

If these efforts are of high enough quality and we agree on the format (a decision which may be influenced by the existence of good documentation in that format), they could presumably be used to jump-start such a migration.

**Comment by nauta-turbidus** [comment]
> I'm aware of such efforts, they're a good thing for sure, but that doesn't mean we shouldn't pursue having a more useful source of documentation truth in the engine?
> 
> If these efforts are of high enough quality and we agree on the format (a decision which may be influenced by the existence of good documentation in that format), they could presumably be used to jump-start such a migration.

I'm obviously saying such efforts may be a good start for something to eventually include into engine itself.

**Comment by corpserot** [comment]
for reference in this issue, here's the efforts ongoing that i know of:
- (me, 5.14) https://github.com/corpserot/luanti_lsp_definitions/
- (archie, 5.14-dev?) https://git.minetest.land/archie/luanti-api/
- (team lord, 5.8?) https://github.com/Voxrame/luanti-ide-helper/
- and [Luanti#16575](https://github.com/luanti-org/luanti/pull/16575) that just started
- there are previous efforts, but they have stalled or no longer have activity.

agree with centralizing efforts in the engine. i'll see if i can draft a mock of that special format. ~~actually scratch that.~~ it's gonna take a while

**Comment by potpiefox** [comment]
I'm the maintainer of [git.minetest.land/archie/luanti-api](https://git.minetest.land/archie/luanti-api/). So far the repo has documented 130 types, 286 core functions, 240 member functions (along with all pertinent documentation for their respective symbols), I've tried to keep as much parity with the original documentation as possible, but it's hard to do without assistance from the dev team. Parts of the API are unclear, ambiguous, and behavior has to be surmised.

We (@corpserot and I) have discussed deriving information from the engine's implementation, but we have reached the conclusion that this design would be backwards. In principle, the API should define the behavior of the engine, not the other way around. [Otherwise we risk breaking things in the future.](https://cwe.mitre.org/data/definitions/758)

I believe we need to clarify the API's behaviors where it's unclear, and only then we can start creating accurate notations. But changing API contracts is serious, and there's a lot that can go wrong if we don't keep backwards compatibility in mind. For that reason, I also believe we need a coordinated effort with the aid of Luanti developers. And frankly, seeing so much parallel, duplicate work being done is discouraging...

**Comment by mark-wiemer** [comment]
Hi everyone, jumping in to say I've been thinking about this for the past few weeks and have been doing my own research as well. I've come to the same conclusions as folks here, I think the canonical representation should be LuaCATS, and we can use LuaLS to generate Markdown, then other custom parsers to generate Teal, TypeScript, etc. as needed. But other langs are far off, for now we need to focus on rewriting the docs in a consistent format, maintaining existing information, building in new important information, and keeping a consistent voice. Please join the luanti-docs Discord / IRC for further discussion if you're interested because this will probably be my project for 2026 :) if I can get approval from other members of the docs team, I can start breaking this work into sub issues and accepting PRs for a new `luanti_api.lua` file that we can use to programmatically generate `lua_api.md`. Ref https://github.com/luanti-org/docs.luanti.org/issues/296 for docs team research.

**Comment by mark-wiemer** [comment]
Is https://git.minetest.land/archie/luanti-api/ working for others? I've been getting a 404 for the past couple weeks, anyone able to share access with me? https://git.minetest.land/mark-wiemer

**Comment by corpserot** [comment]
@mark-wiemer the new place is https://git.minetest.land/andro/luanti-api/
